consensus/hotstuff/verification/combined_verifier_v3.go

```diff
@@ -43,112 +43,118 @@ func NewCombinedVerifierV3(committee hotstuff.Committee, packer hotstuff.Packer)
+func (c *CombinedVerifierV3) VerifyQC(signers flow.IdentityList, sigData []byte, block *model.Block) error {
 
 	dkg, err := c.committee.DKG(block.BlockID)
 	if err != nil {
-		return false, fmt.Errorf("could not get dkg data: %w", err)
+		return fmt.Errorf("could not get dkg data: %w", err)
 	}
 
 	// unpack sig data using packer
 	blockSigData, err := c.packer.Unpack(block.BlockID, signers.NodeIDs(), sigData)
 	if err != nil {
-		return false, fmt.Errorf("could not split signature: %w", modulesig.ErrInvalidFormat)
+		return fmt.Errorf("could not split signature: %w", signature.ErrInvalidFormat)
 	}
 
 	msg := MakeVoteMessage(block.View, block.BlockID)
 
 	// verify the beacon signature first since it is faster to verify (no public key aggregation needed)
 	beaconValid, err := dkg.GroupKey().Verify(blockSigData.ReconstructedRandomBeaconSig, msg, c.beaconHasher)
 	if err != nil {
-		return false, fmt.Errorf("internal error while verifying beacon signature: %w", err)
+		return fmt.Errorf("internal error while verifying beacon signature: %w", err)
 	}
 	if !beaconValid {
-		return false, nil
+		return fmt.Errorf("invalid reconstructed random beacon sig for block (%x): %w", block.BlockID, model.ErrInvalidSignature)
 	}
 
 	// verify the aggregated staking and beacon signatures next (more costly)
 
-	verifyAggregatedSignature := func(pubKeys []crypto.PublicKey, aggregatedSig crypto.Signature, hasher hash.Hasher) (bool, error) {
-		// TODO: to be replaced by module/signature.PublicKeyAggregator in V2
+	verifyAggregatedSignature := func(pubKeys []crypto.PublicKey, aggregatedSig crypto.Signature, hasher hash.Hasher) error {
+		// TODO: to be replaced by model/signature.PublicKeyAggregator in V2
 		aggregatedKey, err := crypto.AggregateBLSPublicKeys(pubKeys)
 		if err != nil {
-			return false, fmt.Errorf("could not compute aggregated key: %w", err)
+			return fmt.Errorf("internal error computing aggregated key: %w", err)
 		}
 		valid, err := aggregatedKey.Verify(aggregatedSig, msg, hasher)
 		if err != nil {
-			return false, fmt.Errorf("internal error while verifying aggregated signature: %w", err)
+			return fmt.Errorf("internal error while verifying aggregated signature: %w", err)
 		}
-		return valid, nil
+
+		if !valid {
+			return fmt.Errorf("invalid aggregated staking sig for block %v: %w", block.BlockID, model.ErrInvalidSignature)
+		}
+
+		return nil
 	}
 
 	// first fetch all beacon signers public keys
 	beaconPubKeys := make([]crypto.PublicKey, 0, len(blockSigData.RandomBeaconSigners))
 	for _, signerID := range blockSigData.RandomBeaconSigners {
 		keyShare, err := dkg.KeyShare(signerID)
 		if err != nil {
-			return false, fmt.Errorf("could not find key share for signer %v: %w", signerID, err)
+			// the packer should have found the signer, otherwise the packer will return an ErrInvalidFormat
+			// and we won't reach it. So reaching here would be an exception
```

Comment:
the packer only checks that the ID corresponds to a valid consensus participant. Not sure what happens, if the consensus participant failed the DKG (?) Will we retrieve a dummy key?

Your comment is correct, as long as we (implicitly) assume that every consensus participant is also a random beacon participant. While we have this convention at the moment, it is theoretically also possible to only have a _sub-set_ of the consensus committee run the Random Beacon. I think it is worth-while to document this assumption here: 

 ```suggestion
// We assume that every consensus member is also a random beacon participant. The 
// Packer checks that only valid consensus participants are in the signers list. Otherwise,
// the packer should return an ErrInvalidFormat. Consequently, we expect for every 
// signer a DKG record. Reaching this failure case indicates a bug in the internal logic.
```

Reply:
>Not sure what happens, if the consensus participant failed the DKG (?) Will we retrieve a dummy key?

@AlexHentschel Let me confirm your question:
Say a consensus participant node B failed the DKG, and we are node A , who successfully completed the DKG. The question is: will node A retrieve a dummy public key for node B? 

My understanding is yes, if DKG has completed, we will have a key for every consensus participant, regardless whether each participant actually succeeded or failed DKG. cc @tarakby to confirm

And your question also reminds me one more thing that I would like to double check. 
There are two events: 
1. DKG completed
2. DKG ended

The difference between these two events are:
- event 1 happens _near_ the end of the current Epoch, whereas event 2 happens _at_ the end of the current Epoch. 
- event 1 will produce locally the private key for myself for the next epoch, whereas event 2 will receive from the block payload the DKG keys for all DKG participants for the next epoch. 
- The DKG keys from event 2 is the canonical DKG keys for all DKG participants, which can be used verify the local genereated DKG private key generated from event 1. 

The question is would event 1 also produce locally the DKG keys for all consensus participants? 
If yes, then is there a possibility that it will be different from the canonical DKG keys from event 2? 
If yes, then we need to ensure that the `dkg.KeyShare` will read the key share from event 2, not event 1. Right? Is this done already? cc @tarakby @jordanschalm

Reply:
> event 1 happens near the end of the current Epoch, whereas event 2 happens at the end of the current Epoch.

Both happen strictly before the end of the current epoch. Event 2 corresponds with the transition to the `EpochCommitted` phase. Then we have the entirety of that phase before the end of the current epoch (typically at least an hour).

Clearer descriptions might be:
Event 1 - local DKG completion (I have my key, and submitted my results)
Event 2 - global DKB completion (I observe the results committed to by a majority of participants)

> The question is would event 1 also produce locally the DKG keys for all consensus participants?

Yes

> If yes, then is there a possibility that it will be different from the canonical DKG keys from event 2?

Yes

> If yes, then we need to ensure that the dkg.KeyShare will read the key share from event 2, not event 1

Yes this is what we do. Only the canonical DKG keys from the service event are incorporated into the protocol state.